docs/proposal-0007-discrete_metric_execution.md

```diff
@@ -0,0 +1,107 @@
+
+Periodically, Jetstream fails during automated analysis of experiments. Often, this is due to resource limits, which occur in either BigQuery or the Jetstream cluster. Each failure requires manual debugging from DE and/or DS to determine the issue and come up with a solution. Often, the required solution is suboptimal and requires substantial compromise. For example, common solutions include removing metrics from the experiment, which leads to less information from the experiment, and downsampling the experiment population, which leads to lower confidence/precision to results.
+
+### Other Considerations
+There are many improvements we would like to make to analysis that could be enabled depending on the approach we take to these Jetstream architecture changes, for example:
+- **Covariate adjustment**
+  - requires some analysis periods to be dependent on others
+  - caveat: preenrollment analysis periods may never be available (e.g., new user experiments)
+- **Shredder support**
+  - allow for easily rerunning specific parts of analysis along with downstream tasks (e.g., rerun a metric and its statistics)
+  - additionally, we could help to prevent loss of fidelity from rerunning an experiment after clients have been shredded by aggregating larger windows using daily aggregations, and running daily aggregations first
+- **Client Filtering**
+  - we want to be able to filter out clients with certain attributes (e.g., duplicate enrollments), and this could potentially be affected by changes to Jetstream's architecture
+
+
+## Decision Drivers
+
+* Ability to scale without concern for number of metrics
+* Reduce manual support needed for automated analysis errors (reduce/eliminate errors)
+* Enable (or not inhibit) known upcoming features
+* Ability to (re-)run individual metrics
+
+### Non-Drivers
+
+* We are not specifically setting **cost reduction** as a goal of these changes. We will, however, be monitoring for any cost changes that result. Additionally, something we do not want is for the proposed solution to include significantly increased costs as an *expectation*.
+* Similarly, we are not setting **total execution time** as a goal of these changes. The scalability decision driver is similar to this, but what we care about is that we can scale without failing, not without increasing total execution time.
+
+
+## Description of Solution
+
+The chosen solution involves a few primary components that will each contribute to some or all of the decision drivers.
+
+### Discrete Metric Execution
+
+For each metric (parallel execution):
+* query metric from telemetry (we can aggregate all metrics for a given data source)
+* write to metric table with results (delete existing **column data** first (currently we delete and re-write the entire table with one query))
+* compute statistics for the metric
+* write to statistics table with results (behavior unchanged)
+
+#### Questions
+* query for every metric or for every metric *data source*?
```

Comment:
I _imagine_ that the per-data-source option would be cheaper (only have to scan each row once?), but maybe more memory intensive?

Reply:
Does the per-data-source option allow us to easily add a new metric from a data source that's already being used for other metrics? or would it require rerunning all the metrics from that data source?

Reply:
>I imagine that the per-data-source option would be cheaper (only have to scan each row once?), but maybe more memory intensive?

Yea this was my thinking as well.

>Does the per-data-source option allow us to easily add a new metric from a data source that's already being used for other metrics? or would it require rerunning all the metrics from that data source?

My plan right now is to run it per-datasource, but allow for a single metric to run individually as well (via a CLI option).